File: src/main/java/at/yawk/math/data/BinaryExpression.java
package at.yawk.math.data;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * @author yawkat
 */
@Getter
@RequiredArgsConstructor
@EqualsAndHashCode
public abstract class BinaryExpression implements Expression {
    private final Expression left;
    private final Expression right;

    protected abstract String toString(String lhs, String rhs);

    @Override
    public String toString(int radix) {
        return toString(left.toString(radix), right.toString(radix));
    }

    @Override
    public String toString() {
        return toString(10);
    }
}

File: src/main/kotlin/at/yawk/math/data/BinaryExpression.kt
package at.yawk.math.data

/**
 * @author yawkat
 */
abstract class BinaryExpression(val left: Expression, val right: Expression) : Expression {
    protected abstract fun toString(lhs: String, rhs: String): String

    override fun toString(radix: Int): String {
        return toString(left.toString(radix), right.toString(radix))
    }

    abstract override fun equals(other: Any?): Boolean
    abstract override fun hashCode(): Int
}